### PR TITLE
Simplify project lifecycle

### PR DESCRIPTION
**Pull request by lehors**
This PR replaces the Dormant, Deprecated, and End of Life states with a single Archived state from which a project can possibly move back to Incubation.

### CODE REVIEW

**Review by tkuhrt — APPROVED** [review]

**Comment by lehors** [comment]
Note that one could argue that this is going too far by removing Deprecated but I think the same functionality can be achieved as described and the gain in simplicity outweighs the loss.

**Comment by petermetz** [comment]
@lehors First of all: LGTM. 

> Note that one could argue that this is going too far by removing Deprecated but I think the same functionality can be achieved as described and the gain in simplicity outweighs the loss.

We could make it part of the guidelines that there has to be an explanation provided at the very top of the README.md about the state of the project that clears up ambiguities related to not having the 3 distinct states available to chose from anymore. 
It could even be a set of pre-defined questions like we have it in the quarterly reports, but a sort of exit interview for the maintainers that would seek to locate the position of the project on the spectrum [1].

I'm thinking of questions like:
1. Would you/your organization be interested in rebooting the project if other people/organizations came along to share the burden of maintenance?
2. Are you archiving the project because of fundamental design flaws that aren't feasible to be fixed?
3. Are you archiving the project because of fundamental security flaws that cannot  be fixed without a complete redesign? (maybe this is the same question as 2)...)

[1] The spectrum being worst to best case in terms of the subjective feelings of the maintainers, e.g. it could be that `a)` they archived the project because of critical design failures/security vulnerabilities that are not feasible to be fixed without a complete rewrite and they'd rather just start over as a lab OR `b)` they are loving the project and they would love to continue working on it but funding has ran out but if someone else were to pick that up the project could chug along nice and healthy.

**Review by petermetz — APPROVED** [review]

**Review by mbrandenburger — APPROVED** [review]
Great simplification! I like it!

File: governing-documents/project-lifecycle.svg
Comment (tkuhrt) [inline governing-documents/project-lifecycle.svg:1]: Two questions that we should consider with this diagram:
1. Can a project go straight to graduated?
2. Can a project move from graduated back to incubation?

Comment (lehors) [reply governing-documents/project-lifecycle.svg:1]: As proposed the answer to these questions is: no (both).
Any reason you think it should be different?

**Review by swcurran — APPROVED** [review]

**Comment by tkuhrt** [comment]
@lehors : I believe these were the options we discussed in the TOC meeting today.

# Option 1 - Arnaud's Version
[](https://mermaid.live/edit#pako:eNpFj8EKwjAQRH8l7Ln9gRwEQRAPguDNxsOapG0gyYY0UaT0302j0tu83RmYmUGS0sCht_SSI8YkPGOhu0QKNKG9r2i6k5f5gcmQr4ehO0ZUGZNWlbHbRzma5w8Da9sdMzVa5bBJrPlN4t8LDTgdHRpV2szrS0AatdMCeJFK95htEiD8UqyYE13fXgJPMesGclClzcHgENEB79FO5RrQ34g21sokiufv4jp8-QAzrFPT)

# Option 2 - Option 1 + Graduated back to Incubation
[](https://mermaid.live/edit#pako:eNo1UE1rwzAM_StG5_YPhDEoDMoOg0Fvs3dQbSUxJJax5Y5R-t8rJ9tJ7wvxeHfwHAgGGBf-8TMWccmYbD8LZ664fHda23UqmGeD1Z68xBuZi-BEdXONifY9-XZFiZz-pMmeC4aGQmFTKIV-0J6Kn_XBrmZzPL7q146jeelk2mJ1N1yCA6xUVoxBO96750BmWsnBoDDQiG0RBy49NIpN-PKbPAxSGh2g5aAN3iJq_fVfzJi-mJWOuFTlFKJw-dhn2NZ4PAH-QVz6)

# Option 3 - Option 1 + Pre-Archive State
[](https://mermaid.live/edit#pako:eNpVkMEKwjAQRH-l7Ln9gRwEQRAPguDNxsOabNtAmw1pokjpv5vGQvX2ZnZ3GHYCxZpAQNPzS3Xog7RF4eqLZ8cj9vdFmvpkVXxgMGyz0dZHjzpiIJ21w3RA1d6rzjxXD-s_6Yqq2hUmx2VsN3SYQ3_Y4ba-cvKhhIH8gEanwtMykxA6GkiCSKipwdgHCdLOaRVj4OvbKhDBRyohOp0KHwy2HgcQDfZjch3aG_OmSZvA_vx9Sv7N_AHI9F9E)

# Option 4 - Option 2 + Pre-Archive State
[](https://mermaid.live/edit#pako:eNpVkM1qwzAMx1_F6Ny8QBiDwqD0MBj0NnsH1VYSQ2IbW94Ype8-OS6UnaT_T9-6gY2OYIRpjT92wcwmKJX0R44pFly_miz1OmdMi8Kij5b9N6kL40xljyrl9TnYekX2MTzQrE8ZXUUmtxMKrhnUx2wXadBpQhlEw3-mhuFVJjXfq5cm5r20dJGwV_Y0E-AAG-UNvZMrbi1mgBfayMAorqMJ68oGTLhLKlaOl99gYeRc6QA1OdnxzaMcuME44VqEJgyfMT41Oc8xv_dP7Q-7_wFummgS)

# Option 5 - Radical
[](https://mermaid.live/edit#pako:eNpNj0EKwjAQRa9SZt1eIAtBcCsI7mxcDMnERpJMSBNFSu9uGgXd_fffwPAXUKwJBBjHTzVhyjJ0XRxPiSPP6K4bVriTagbHfVKTfZBuJnbDsPv339hqhB48JY9W1wfLpiXkiTxJEDVqMlhcliDDWk-xZD6_ggKRU6EeStSY6WDxltCDMOjm2kYMF-Yfk7aZ0_Ezom1Z39CdR80)

(*) Note in this option, we would need to implement badges.

**Comment by swcurran** [comment]
Awesome work, Tracy — thanks.  During the meeting I had written in my notebook pretty much exactly version 4, so that’s my choice, with the rational for the “pre-archived” (need a better name! 😄 ) state what I said in the meeting.

**Comment by tkuhrt** [comment]
> “pre-archived” (need a better name! 😄 )

Naming is hard. Open to suggestions.

**Comment by swcurran** [comment]
> > “pre-archived” (need a better name! 😄 )
> 
> Naming is hard. Open to suggestions.

My suggestions, in order:

“Inactive”, “Dormant”, “Unsponsored”, “Unmaintained”, “Stagnant”

**Comment by lehors** [comment]
Thanks for doing all the work @tkuhrt ! :-)
I don't really know how to interpret the "active stages" box in 2 and 4. Does this mean a proposal can go directly to graduated?
For option 5, shouldn't there be a path from Archived back to Project (or Proposal)?

**Comment by swcurran** [comment]
> For option 5, shouldn't there be a path from Archived back to Project (or Proposal)?

I would say it just makes the picture messy to include that. Someone *could* do that — it’s open source code after all — but I don’t think it’s worth calling out.

**Comment by tkuhrt** [comment]
> I don't really know how to interpret the "active stages" box in 2 and 4. Does this mean a proposal can go directly to graduated?

Oh...this was one of my earlier questions that we did not discuss during the meeting. Yes, this would imply that a proposal could go directly to graduated if it met the incubation exit criteria already.

If you don't like that, then there is always:

# Option 6

[](https://mermaid.live/edit#pako:eNpFj8EKwjAQRH8l7Ln9gSCCIIgHQfBm42FN0jaQZkOaKFL676ap0tub2dllZwJJSgOH1tJb9hiicIz55hrI04j2sUjTnJ1MT4yGXDG65hRQJYxaFY3NIcjevH7Ss7reM1NW2W7hbuUFsRzYEP9hqGDQYUCj8jvTMhIQez1oATyj0i0mGwUIN-copki3j5PAY0i6guRVfudosAs4AG_Rjtn16O5Em9bKRAqXtXJpPn8BerdUDw)

# Option 7

[](https://mermaid.live/edit#pako:eNpVkMEKwjAQRH8l7Nn-QBFBEMSDIHgz8bAm2zbQZkOaKFL676ZRUG9vZneHYSfQbAhqaHp-6A5DVE4IL0-BPY_YXxdp5cHpdMNo2RWjlfuAJmEkU7THfEDVNujO3j8eyj_pRVVthC1xYr1w--YFPZbUH_b43f9w9mEFA4UBrcmNp2WmIHY0kII6o6EGUx8VKDfnVUyRz0-noY4h0QqSN7nxzmIbcIC6wX7Mrkd3Yf5qMjZyOL6_Up4zvwAan1-A)

**Comment by tkuhrt** [comment]
> For option 5, shouldn't there be a path from Archived back to Project (or Proposal)?

For this, I think it would go back to proposal. So, I guess I agree with @swcurran on leaving it out as we don't really have the entry arrow into proposal either.

**Comment by tkuhrt** [comment]
> Thanks for doing all the work @tkuhrt ! :-)

Hope you don't mind me jumping in. If anyone thinks there are other options that should be considered, you can click on the picture that you want to build on, and it should take you to mermaid.live where you can edit the picture, copy the markdown into a comment and show your preferred option.

**Comment by petermetz** [comment]
My preference at this moment (might change later!): option 5 plus badges, but we need to have the badges nailed down first for that to work IMO.
If the badges are operational and satisfactory (happy path) then we could move to option 5, but otherwise option 6.

**Comment by lehors** [comment]
> > I don't really know how to interpret the "active stages" box in 2 and 4. Does this mean a proposal can go directly to graduated?
> 
> Oh...this was one of my earlier questions that we did not discuss during the meeting. Yes, this would imply that a proposal could go directly to graduated if it met the incubation exit criteria already.

I have to admit to be hesitant about this change. Our position has always been that Incubation was the way for any project (even mature ones) to get into the Hyperledger organization and that if a project was mature enough it could quickly move to Gratudated status after that but that Incubation was a necessary first step to establish itself as being part of Hyperledger.
Of course the TOC can change its position on this.

**Comment by denyeart** [comment]
My vote is Option 3 and call pre-archived "Dormant". This leaves things pretty consistent with the existing lifecycle but collapses "Deprecated" into "Dormant" to simplify it.

Option 2 and 4 are good in theory but the diagram hurts my brain, and would require constant TOC re-assessment between Incubated and Graduated.

In my opinion Option 5 "Radical" is too radical. Badges seem too nuanced for people to understand at a quick glance. Whereas "Incubated" to "Graduated" provides intuitive meaning at a glance and a good incentive for new projects to meet the criteria and make the leap.  I'm not against badges per se, perhaps once we define the badges well we could say that once a project obtains badges X, Y, Z then they are automatically Graduated. But I don't think we have to decide that for now.

**Comment by swcurran** [comment]
Noting that Options 3 and 7 are the same. In the Discord Poll I voted for 7, so also 3.  :-)

**Comment by tkuhrt** [comment]
@swcurran : The difference between 3 and 7 is that 7 allows projects to move backwards to incubation from graduated.

**Comment by tkuhrt** [comment]
> My vote is Option 3 and call pre-archived "Dormant". This leaves things pretty consistent with the existing lifecycle but collapses "Deprecated" into "Dormant" to simplify it.

My problem with "Dormant" state (regardless of what we call it) is that projects that have gone dormant have never come back from dormant and in some cases, it is extremely hard to contact the maintainers after it goes into that state. If we have a tool that will automatically move a project to dormant on 3 months of inactivity by the maintainers and then to archived after another 3 months of inactivity, I might be willing to have the extra state. Right now, this state seems like noise. As an example, currently, Transact has been Dormant (officially) since May 4, 2023; however, it was Dormant (unofficially) since before that time. What is the right time to leave the repository un-archived before we finally make the decision to archive it?

**Comment by swcurran** [comment]
> My problem with "Dormant" state (regardless of what we call it) is that projects that have gone dormant have never come back from dormant and in some cases, it is extremely hard to contact the maintainers after it goes into that state. If we have a tool that will automatically move a project to dormant on 3 months of inactivity by the maintainers and then to archived after another 3 months of inactivity, I might be willing to have the extra state. Right now, this state seems like noise. As an example, currently, Transact has been Dormant (officially) since May 4, 2023; however, it was Dormant (unofficially) since before that time. What is the right time to leave the repository un-archived before we finally make the decision to archive it?

Having been through this with Ursa, I definitely disagree with this.  The goal of the state is to notify a members of the community that might want alter the direction to take action.  In some cases, they will not, and the transition to dormant will follow on to archived without intervention.  However, we leave open the opportunity for two things to happen by those not aware of the current state of maintainers in a project:

- Collaborators might step up and take over the project, keeping it alive. Such collaborators might have no idea that the project is need of maintainers.
- Collaborators might step up and get what is useful out of the project moved elsewhere before the official archiving.

The latter would have made the Ursa transition much easier for the community, vs. the messy scramble it became. 

I have no problem with a time being set for the transition from Dormant to Archived if nothing changes.  That can be set at the time the transition to dormant is made — perhaps a range from 1 week to 3 months. If there is no value in keeping the project, make it super short. If the project is in active use, make it a marketing opportunity to see if there is interest — and to enable a graceful shutdown.

**Comment by lehors** [comment]
> Having been through this with Ursa, I definitely disagree with this. The goal of the state is to notify a members of the community that might want alter the direction to take action. In some cases, they will not, and the transition to dormant will follow on to archived without intervention. However, we leave open the opportunity for two things to happen by those not aware of the current state of maintainers in a project:
> 
>     * Collaborators might step up and take over the project, keeping it alive. Such collaborators might have no idea that the project is need of maintainers.
> 
>     * Collaborators might step up and get what is useful out of the project moved elsewhere before the official archiving.

But if that were the case a project can easily be unarchived. This is something that has happened several times in the labs. A lab appears dormant it gets archived and then someone comes out of the woodwork and says "hey, I would actually like to work on this". The lab is unarchived (before the ink is dry in typical @ryjones fashion :-) and the work resumes. No foul.

I think the change from End Of Life to Archived is the reason we don't need any "dormant/pre-archived" state.

Admittedly you might think that archiving a project is quite a harsh way to let the community know they really need to step up if they want the project to go on but it greatly simplifies the management of those projects that otherwise stay in limbo for long period of times.

**Comment by denyeart** [comment]
In addition to what @swcurran said about alerting members of the community, I think the other purpose of a pre-archived Dormant state is to allow for final pull requests before a project is archived. Even though a project may be generally inactive, it doesn't mean the pull request rate will go straight to zero. Some example pull requests against a Dormant project:

- Maybe somebody wants to update the Readme with transition plans and options as they become available.
- Maybe somebody wants to update a project dependency to fix a vulnerability for remaining users.
- Maybe an external change has caused CI to break and somebody wants to fix it.

So even though the codebase is generally inactive and not moving forward, it doesn't mean we should immediately block all pull requests which is what the archived state would imply. We could state that a Dormant project automatically goes to Archived after N months of no activity, so that TOC doesn't have to re-visit the project decision yet again.

Ultimately I think we will be quicker to mark a project Dormant than we would to completely archive it, giving the community an important signal sooner than we otherwise would that the project is winding down and now is the time for any final pull requests or to step up as new maintainers.

**Comment by swcurran** [comment]
> Admittedly you might think that archiving a project is quite a harsh way to let the community know they really need to step up if they want the project to go on ...

That’s exactly what I’m thinking, and I think it is worth the extra bit of overhead to go through the process in two steps instead of one.

**Comment by VRamakrishna** [comment]
Sorry for my late entry to the conversation. Can we redraw option 2 so that all stages are vertically stacked but which consists of an Incubation-Graduation cycle?

**Comment by VRamakrishna** [comment]
Option 4 is ambiguous, as the back arrow from Pre-Archived may lead to Incubated or Graduated. I would replace it with "Option 7 + back-arrow from Pre-Archived to Graduated".

**Comment by tkuhrt** [comment]
> Sorry for my late entry to the conversation. Can we redraw option 2 so that all stages are vertically stacked but which consists of an Incubation-Graduation cycle?

[](https://mermaid.live/edit#pako:eNpFkMFqwzAMhl_F6Ny-QBiDjsHYYTBoT7N3UC0lMSS2keWNUfrus5PBTtL__T-yrBv4RAwDjEv69jOKumhMtu-Sciq4fHZZ6nUSzLPBYk9ewxebs-LEZXONoSDccIrm8rSTYF-jr1fs8C802RdBqqhMG-FIvaA9iZ_byJ1mczw-tnd6H8xDF9MWK7vhIhxgZVkxUNv61j0HOvPKDobWEo9YF3Xg4r1FsWo6_0QPg0rlA9RMbYPngO1DKwwjLqXRjPEjpX_NFDTJ236Z7UD3XwYkYwY)

**Comment by tkuhrt** [comment]
Based on Hart's comments in the TOC meeting:
[](https://mermaid.live/edit#pako:eNpVkE0KwjAQRq8SZm0vUEQQBHEhCO5sXIzJtA20mZAmipTe3TQVrLv3vvlhmBEUa4IS6o5fqkUfpBXCVRfPjgfs7rOa6mRVfGAwbHPQVEePOmIgnd1hGqBi71Vrnt8Mqz91oih2wuR1Yjtzs_CMDvPWFTtcZMUphw305Hs0Ol08zjUJoaWeJJQJNdUYuyBB2im1Ygx8fVsFZfCRNhCdThcfDDYeeyhr7IaUOrQ35p-TNoH9eflKfs70AR5JX4c)

**Comment by tkuhrt** [comment]
[](https://mermaid.live/edit#pako:eNpNj80KwjAQhF8l7Nm-QBBBKIgHT95MPKxJ2gaaH-JGkdJ3N00Fe_tmZhl2JlBBG-DQjeGtBkwkPWNWnL3KDyQb_H0xenFKqDOS0VVHFG1IDj1VieKY1GBfv9SyfdMcWL_yghFry4YjVmE3XHzYgTOl1-ry0rRkEmgwzkjgBbXpMI8kQfq5nGKmcP14BZxSNjvIUZcPW4t9Qge8w_FZ3Ij-FsJfG20ppMs6u66fv-jUVZ4)

**Comment by tkuhrt** [comment]
[](https://mermaid.live/edit#pako:eNpNkM0KwjAQhF8l7Ll9gSCCUBAPnrzZeFiTtA00P8SNIqXvbpqK9vbNzLI77ATSKw0cutG_5ICRhGPMtCcn0x3JeHdbjL49RlQJSauiA7aNjxYdFYntIcrBPL-pYbu63rN-5QUDli0bDliE2XDxfzZUYHU-YVRuNy2RABq01QJ4RqU7TCMJEG7Oo5jIX95OAqeYdAUpqFy2MdhHtMA7HB_ZDeiu3v-1VoZ8PK8fKI-YPyY4WEo)

**Comment by vaporos** [comment]
It would be helpful to clarify what resources are immediately/eventually removed from the project upon changing states. For example, presumably Archived means to turn the github repos read-only and remove the chat channels (offering no real hope of a project's resurrection). We experienced this with Grid, where there is now no location to even talk about it.

I'd also like to see a "Moved" state, which indicates the project left Hyperledger (presumably because the project expectations couldn't be met, but the project remains viable in a different setting). Project could be de-branded Hyperledger and links could be made to the new location.

**Review by tkuhrt — APPROVED** [review]

**Review by petermetz — APPROVED** [review]

**Comment by lehors** [comment]
> It would be helpful to clarify what resources are immediately/eventually removed from the project upon changing states. For example, presumably Archived means to turn the github repos read-only and remove the chat channels (offering no real hope of a project's resurrection). We experienced this with Grid, where there is now no location to even talk about it.

I agree that's unfortunate. I think upon request the chat channel could be left open for a while to address this problem. This doesn't really cost anything.

> I'd also like to see a "Moved" state, which indicates the project left Hyperledger (presumably because the project expectations couldn't be met, but the project remains viable in a different setting). Project could be de-branded Hyperledger and links could be made to the new location.

I don't think this warrants adding another state. There are different reasons for which a project can be archived and that's just one of them. When a project is archived its README should be updated to provide appropriate information for people to know what happened and in case the project has moved elsewhere a pointer to it can be provided.

**Comment by tkuhrt** [comment]
Since the TOC voted and approved this simplification at the last meeting, are we good to merge this? @hyperledger/toc

**Review by swcurran — APPROVED** [review]

**Pull request merged**